File: src/test/java/com/github/t1/yaml/parser/Spec.java
package com.github.t1.yaml.parser;

import com.github.t1.yaml.parser.Expression.ReferenceExpression;
import lombok.NonNull;
import lombok.Value;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

class Spec {
    final List<Production> productions;
    private final Map<String, Production> index = new TreeMap<>();

    Spec(List<Production> productions) {
        this.productions = productions;
        index();
    }

    private void index() {
        for (Production production : productions) {
            Production previous = index.put(production.getKey(), production);
            assert previous == null : "overwrite " + previous.getKey();
        }
        for (Production production : productions) {
            production.expression.guide(new Expression.Visitor() {
                @Override void visit(ReferenceExpression referenceExpression) {
                    String ref = referenceExpression.ref;
                    production.references.put(ref, index.get(ref));
                }
            });
        }
    }

    Production get(String key) {
        Production result = index.get(key);
        if (result == null)
            throw new RuntimeException("no production '" + key + "' found");
        return result;
    }

    @Value static class Production {
        public int counter;
        public @NonNull String name;
        public String args;

        public @NonNull Expression expression;

        private final Map<String, Production> references = new HashMap<>();

        String getKey() { return name + ((args == null) ? "" : "(" + args + ")"); }

        @Override public String toString() {
            return "[" + counter + "] : " + name + ((args == null) ? "" : " [" + args + "]") + ":\n"
                + "  " + expression;
        }
    }
}

File: src/test/java/spec/generator/Spec.kt
package spec.generator

import spec.generator.Expression.ReferenceExpression
import java.util.TreeMap

class Spec(val productions: List<Production>) {
    val externalRefs = mutableSetOf<String>()
    private val index = TreeMap<String, Production>()

    init {
        index()
        resolveReferences()
    }

    private fun index() {
        for (production in productions) {
            val previous = index.put(production.key, production)
            assert(previous == null) { "overwrite " + previous!!.key }
        }
    }

    private fun resolveReferences() {
        for (production in productions) {
            production.expression.guide(object : OrIgnoreVisitor() {
                override fun visit(reference: ReferenceExpression) {
                    val target = index[reference.key]
                    if (target == null) externalRefs += reference.key
                    else production.addReference(reference, target)
                }
            })
        }
    }

    operator fun get(key: String): Production = index[key]
        ?: throw RuntimeException("no production '$key' found")
}

data class Production(
    val counter: Int,
    val name: String,
    val args: List<String>,
    val expression: Expression) {

    val referencedProductions = mutableMapOf<String, Production>()
    val referencedExpressions = mutableSetOf<ReferenceExpression>()

    fun addReference(reference: ReferenceExpression, target: Production) {
        referencedExpressions += reference
        referencedProductions[reference.key] = target
    }

    val hasArgs: Boolean get() = args.isNotEmpty()
    val key: String get() = name + argsKey
    val argsKey: String get() = if (args.isEmpty()) "" else args.joinToString(",", "(", ")")

    override fun toString() = "`$counter` : $key:\n$expression"
}
